File: backend/Controllers/DashboardController.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using StockwellApi.Data;

namespace StockwellApi.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class DashboardController : ControllerBase
{
    private readonly ApplicationDbContext _db;

    public DashboardController(ApplicationDbContext db) => _db = db;

    private Guid? UserId => Guid.TryParse(User.FindFirstValue(ClaimTypes.NameIdentifier) ?? User.FindFirstValue("sub"), out var id) ? id : null;

    private async Task<bool> IsAdminAsync() =>
        await _db.UserRoles.AnyAsync(ur => ur.UserId == UserId && (ur.Role == "admin" || ur.Role == "super_admin"));

    [HttpGet("stats")]
    public async Task<ActionResult<DashboardStatsResponse>> GetStats(CancellationToken ct)
    {
        var activeProjects = await _db.Projects.CountAsync(p => p.Status == "active", ct);
        var totalSkus = await _db.Skus.CountAsync(s => s.IsActive, ct);
        var activeOrders = await _db.Orders
            .Where(o => new[] { "for_approval", "approved", "submitted", "delivered", "preparing", "in_transit", "on_hold" }.Contains(o.Status))
            .Join(_db.Projects.Where(p => p.Status == "active"), o => o.ProjectId, p => p.Id, (o, p) => o)
            .CountAsync(ct);
        var activeMembers = 0;
        if (await IsAdminAsync())
            activeMembers = await _db.Profiles.CountAsync(p => p.IsActive, ct);

        return Ok(new DashboardStatsResponse(activeProjects, totalSkus, activeOrders, activeMembers));
    }

    [HttpGet("recent-orders")]
    public async Task<ActionResult<List<RecentOrderResponse>>> GetRecentOrders([FromQuery] int limit = 10, CancellationToken ct=default)
    {
        var orders = await _db.Orders
            .AsNoTracking()
            .Include(o => o.Project)
            .OrderByDescending(o => o.CreatedAt)
            .Take(limit)
            .ToListAsync(ct);

        var result = new List<RecentOrderResponse>();
        foreach (var o in orders)
        {
            if (UserId == null) break;
            var hasAccess = await _db.UserRoles.AnyAsync(ur => ur.UserId == UserId && (ur.Role == "admin" || ur.Role == "super_admin"), ct)
                || await _db.ProjectMembers.AnyAsync(pm => pm.ProjectId == o.ProjectId && pm.UserId == UserId, ct);
            if (!hasAccess) continue;
            result.Add(new RecentOrderResponse(
                o.Id, o.OrderNumber, o.Project?.Name, o.Status, o.SupplierName, o.TotalAmount, o.CreatedAt
            ));
        }
        return Ok(result);
    }
}

public record DashboardStatsResponse(int ActiveProjects, int TotalSkus, int ActiveOrders, int ActiveMembers);
public record RecentOrderResponse(Guid Id, string OrderNumber, string? ProjectName, string Status, string? SupplierName, decimal? TotalAmount, DateTimeOffset CreatedAt);

File: backend/src/controllers/dashboardController.ts
import { Request, Response } from 'express';
import { query } from '../config/database';
import { ApiResponse } from '../dto/common.dto';

export class DashboardController {
  async getStats(req: Request, res: Response) {
    try {
      // active projects
      const projRes = await query('SELECT COUNT(*) as total FROM projects WHERE status = @status', { status: 'active' });
      const activeProjects = (projRes[0] as any)?.total || 0;

      // total skus
      const skuRes = await query('SELECT COUNT(*) as total FROM skus WHERE is_active = 1');
      const totalSkus = (skuRes[0] as any)?.total || 0;

      // active orders (joined with active projects)
      const orderStatuses = [
        'for_approval', 'approved', 'submitted', 'delivered', 'preparing', 'in_transit', 'on_hold'
      ];
      const statusParams = orderStatuses.map((s, i) => `@s${i}`).join(', ');
      const params: Record<string, any> = {};
      orderStatuses.forEach((s, i) => (params[`s${i}`] = s));

      const activeOrdersSql = `
        SELECT COUNT(*) as total FROM orders o
        JOIN projects p ON p.id = o.project_id
        WHERE o.status IN (${statusParams}) AND p.status = @projStatus
      `;
      params.projStatus = 'active';
      const activeOrdersRes = await query(activeOrdersSql, params);
      const activeOrders = (activeOrdersRes[0] as any)?.total || 0;

      // recent orders
      const recentOrdersRes = await query(`
        SELECT TOP 10 o.id, o.order_number, o.status, o.supplier_name, o.total_amount, o.created_at,
               p.id as project_id, p.name as project_name
        FROM orders o
        LEFT JOIN projects p ON p.id = o.project_id
        ORDER BY o.created_at DESC
      `);
      const recentOrders = (recentOrdersRes || []).map((r: any) => ({
        id: r.id,
        order_number: r.order_number,
        status: r.status,
        supplier_name: r.supplier_name,
        total_amount: r.total_amount,
        created_at: r.created_at,
        project: { id: r.project_id, name: r.project_name },
      }));

      // orders by status
      const byStatusRes = await query('SELECT status, COUNT(*) as count FROM orders GROUP BY status');
      const ordersByStatus = (byStatusRes || []).map((r: any) => ({ status: r.status, count: r.count }));

      // active members
      const membersRes = await query('SELECT COUNT(*) as total FROM users WHERE is_active = 1');
      const activeMembers = (membersRes[0] as any)?.total || 0;

      return res.json(new ApiResponse(true, {
        activeProjects,
        totalSkus,
        activeOrders,
        activeMembers,
        recentOrders,
        ordersByStatus,
      }, 'Dashboard stats retrieved'));
    } catch (error) {
      return res.status(500).json(new ApiResponse(false, null, 'Error retrieving dashboard stats', error));
    }
  }
}

export const dashboardController = new DashboardController();
